rust/src/types/amount.rs

```diff
@@ -0,0 +1,582 @@
+use std::{
+    cmp::Ordering,
+    fmt::{Debug, Display},
+    mem,
+    ops::{
+        Add, AddAssign, BitAnd, BitOr, BitXor, Deref, DerefMut, Div, DivAssign, Mul, MulAssign,
+        Not, Rem, RemAssign, Shl, Shr, Sub, SubAssign,
+    },
+};
+
+use bnum::{
+    types::{I256, U256, U512},
+    BInt, BUint,
+};
+
+/// This represents a 256-bit integer. Internally it is a 32 byte array of [`u8`] in big endian.
+#[allow(non_camel_case_types)]
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct u256([u8; 32]);
+
+impl Deref for u256 {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for u256 {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Display for u256 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("0x")?;
+        for (i, byte) in (*self).into_iter().enumerate() {
+            f.write_fmt(format_args!("{:02x}", byte))?;
+            if i % 8 == 7 && i < 31 {
+                f.write_str("_")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Debug for u256 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl u256 {
+    pub const ZERO: Self = Self([0; 32]);
+    pub const MAX: Self = Self([0xff; 32]);
+}
+
+impl From<U256> for u256 {
+    fn from(value: U256) -> Self {
+        // TODO bnum has to_be_bytes with feature nightly
+        let be_value = value.to_be();
+        let bytes: [u8; 32] = unsafe {
+            // U256 = BUint<4>
+            // BUint<4> is transparent wrapper around [u64; 4]
+            mem::transmute(be_value)
+        };
+        bytes.into()
+    }
+}
+
+impl From<u256> for U256 {
+    fn from(value: u256) -> Self {
+        BUint::from_be_slice(value.deref()).unwrap()
+    }
+}
+
+impl From<I256> for u256 {
```

Comment:
Is guess this ignores negative values and just interpreting the 256 bit s of the `I256` as a `U256`? Is this intended or should there be a special case handling?

Reply:
Yes, this is correct and also intended.